### DISCUSSION TITLE
Integration of Spotify and other Content / Services

### DISCUSSION THREAD

**Discussion by MiczFlor** [MiczFlor/RPi-Jukebox-RFID#1835] [Ideas]
We are also curious to learn about alternative services we can connect with and which you would like to see supported -- or have developed support for already.

Spotify has [disabled access to libspotify on May 16, 2022](https://developer.spotify.com/community/news/2022/04/12/libspotify-sunset/). This means Phoniebox can not communicate with Spotify via libspotify (and mopidy-spotify) anymore. The problem is not our code but the cut off by Spotify.

We want Phoniebox users to be able to connect their box to their Spotify accounts and play their content. The possibilities Spotify offers are -- at first glance -- not supporting an integration with Phoniebox. For a potential Spotify re-integration, we are committed to a Spotify ToS compliant way of doing so (both in Version 2 and Version 3).

**Comment by olahelland** [comment]
my kids have been devastated since this stopped working.. they use their box every day. but no more. is there any way of speeding up a new integration? if I knew how to code, I would help out!

**Comment by Toqqi** [comment]
I have just started to test a combination of _raspotify_ (Spotify Connect client based on librespot) and _Spotipy_ (lightweight Python library for the Spotify Web API).
My approach is to enable the Phoniebox as a Spotify Device with _raspotify_ and then control what is played on it by _Spotipy_.
After getting into the documentation at first, I got as far as playing back a Spotify track with the Phoniebox as the target from a simple Python script.
I played around with a few commands and was able to e.g. list all of my private playlists and see all my devices connected to Spotify.
There is an initial Spotify authentication to get the rights from the Spotify user, but I'm sure this is the same as for the implementation with mopidy-spotify.
I'm not 100% sure, if I'm headed into the right direction with this, so any input / feedback is appreciated.

**Reply by olahelland** [reply]
Well, the need for our use is basically just to play a given song, album or playlist. no need for anything else really. play and pause.

**Reply by sonicdee** [reply]
yeah, raspotify & Spotipy work indeed !! as I commented here https://github.com/MiczFlor/RPi-Jukebox-RFID/issues/1815#issuecomment-1136405418 & basic working code is in the gist also, I have two boxes running now stable! but only since I start raspotify every morning again, code is a quick friggle, and has it's downsides (also I'm new to python) - but my kids wanted to have them back asap ;). Setup is a mess, as I needed to create a spotify app, login on a browser capable system (win & copy over .cache file works) ... ^^ I think authentication must be there a kind of easier in/with those projects, but haven't found it fast ...

**Reply by MiczFlor** [reply]
Hi @sonicdee and hi @Toqqi,

the raspotify & Spotipy combination is one of the ideas to overcome the problem of [Spotify dropping their API](https://developer.spotify.com/community/news/2022/04/12/libspotify-sunset/). We are really interested in integrating the [Spotipy](https://spotipy.readthedocs.io/en/2.19.0/) part into Phoniebox, bc this might be the fastest way to get Spotify back on track. Possibly you could provide code and pull requests in the near future to make that happen!?

The big potential **showstopper for a combination of raspotify & Spotipy** is `raspotify`, which uses the [librespot Open Source Spotify client library](https://github.com/librespot-org/librespot). The "[Spotify Developer Terms](https://developer.spotify.com/terms/)", *Section IV Restrictions > 2. General restrictions > a. Misuse of the Spotify Platform > i.* explicitly names **reverse-engineering** as a misuse. We are currently unsure if `librespot` could get us into big trouble -- because of reverse-engineering.

We don't know. But we are trying to figure this out -- and other ways to integrate Spotify again. Phoniebox has been around for over 5 years and gone through many transformations. We don't want to risk what the community of over 100 contributors has achieved in those years.

**Reply by Toqqi** [reply]
> We are currently unsure if `librespot` could get us into big trouble -- because of reverse-engineering.

@MiczFlor 
I was looking also into [spotifyd](https://github.com/Spotifyd/spotifyd) in order to create a Spotify Connect device, but it falls into the same category as librespot, since it reverse-engineers Spotify, unfortunately.

The next thing for me would be to explore the official [Spotify Web Playback SDK](https://developer.spotify.com/documentation/web-playback-sdk/) in order to turn the Phoniebox into a Spotify Connect device. It might be possible to use e.g. nodejs on the Raspberry Pi to run the Javascript code supplied by the SDK.

This might be the cleanest and most compliant solution to reintegrate Spotify.

**Comment by ghost** [comment]
> We want Phoniebox users to be able to connect their box to their Spotify accounts and play their content. The possibilities Spotify offers are -- at first glance -- not supporting an integration with Phoniebox. For a potential Spotify re-integration, we are committed to a Spotify ToS compliant way of doing so (both in Version 2 and Version 3).

Thank you @MiczFlor for those clear words. I fully and entirely agree!

**Comment by Micharius** [comment]
Hello there. 

I tried the solution from user Maahn in the [Issue discussion](https://github.com/MiczFlor/RPi-Jukebox-RFID/issues/1815#issuecomment-1133762447).
I made it quite far in the installation process ([see his github page for the PiCorePhonie project](https://github.com/maahn/piCorePhonie)) until I realized, that this project only supports rc522 card readers, and I have an USB one. At this point, due the lack of knowledge, I gave up.

I downloaded all the content from my daughter's cards (about 50 pieces) as MP3 files. She was just so sad (and unpatient, you know how children can be...) to have her box not working anymore.
I used a quite comfortable Tool for that called [AudFree SpoDable](https://www.audfree.com/spodable/), it costs about €15.  You can just add all the albums in a row and leave the PC running over night or so. I then used an excel file, where I dropped the Phoniebox Card numbers (using the CSV export from Phoniebox) and the MP3 folder name in a table row and some VBA to create the folders on the Phoniebox. That speeded the whole thing a little up. All that took me some time (aka nights) but know, at least, my daughter can use her box again. Fortunately, AudFree pulls all the tags from Spotify and writes it in the MP3 file, including cover image.

Would it be possible to enhance the PiCorePhonie projects, so that it will support USB card readers as well? piCorePhonie uses Python as well, may some codeof the Phoniebox coudl be reused in the PiCorePhonie project.

Kind regards

Micharius

**Comment by rmk-ch** [comment]
Hi, 
I have described two quick workarounds (download as MP3 or switch to TIDAL): https://github.com/MiczFlor/RPi-Jukebox-RFID/issues/1815#issuecomment-1145297487
Both do not require any coding. Hope this helps somebody.
Best,
rmk

**Reply by MiczFlor** [reply]
I added your description to the wiki. Is that ok the way it is? Have a look:
https://github.com/MiczFlor/RPi-Jukebox-RFID/wiki/Integrate-Service-TIDAL

**Reply by rmk-ch** [reply]
This is perfect, thanks.

**Comment by varac** [comment]
I'd like a future(3)-proof solution based on mopidy. Mopidy is great because it supports multiple backends and other backends like modidy-tidal could be plugged in easily as @rmk-ch descirbed above. If this means spotiy won't work but Tidal i.e. I'd be happy to migrate to Tidal (I am using a free 30 day trail and most of the music/audio books I was using on Spotify I also found on Tidal).

From what I can see from the Spotify discussion is that an easy Phoniebox integration would be hard and it would be great to be open to other backends as well.

**Reply by MiczFlor** [reply]
Hi @rmk-ch @varac 
I totally agree that Phoniebox should become modular so that it's open to other services and remain  agnostic to commercial content providers. The TIDAL integration via `mopidy` is a good example for openness regarding services. How could that become a module in Phoniebox? Currently the install script favours Spotify as the only content provider to interface with out of the box.

**Reply by rmk-ch** [reply]
@MiczFlor i am not deeply familiar with the future version. What I like about v2 though is the mopidy backend and its modularity. Why not keep letting mopidy do this job? What is currently not obvious is that a tidal stream is handled the same way as a tidal stream in Phonieboxand mopidy. It is named spotify as well. I think required changes would be purely cosmetic except of maybe an automatic link conversion (e.g. from https://tidal.com/artist/5462037 to tidal:artist:5462037).
This is at least from a deployment view. What I am unsure about is the install script.

**Comment by s-martin** [comment]
Current development status of mopidy-spotify:

https://github.com/mopidy/mopidy-spotify/issues/110

**Comment by MrStrategy** [comment]
I've reviewed the whole thread in [mopidy-spotify/issues#110](https://github.com/mopidy/mopidy-spotify/issues/110) and build my own workflow to apply all the development effort.  

Considering [mopidy-spotify/issues#110](https://github.com/mopidy/mopidy-spotify/issues/110) is still under development I tried to establish the current implementation in one of my phonie boxes. And I finally ended up with the ability to play spotify content again.

**Play** and **Pause** is fully working but there is no chance for fast forward / rewind yet. Also you can use Iris but moving the slider to jump within the track has no effect. But my main use case is fully working: Let my 2 years old son play his own songs.

Some detailed guide how I implemented the solution is available here #1815

**Reply by emutier** [reply]
hi, 
I have been following the breadcrumbs and there have been some very recent comments: 

https://github.com/mopidy/mopidy-spotify/issues/110#issuecomment-1329388690

@MiczFlor: is this of any help to fix the current issue with spotify? 

I am no expert so I hope this information helps those who understand more...

**Reply by helge1979** [reply]
@MiczFlor : Are there any news? do you think a spotify integration possible?

**Comment by Sordit** [comment]
Hi all,

I build a PhonieBox in 2020 for my niece and she was very sad when it stopped working.

Now I will become a daddy myself and as I did still not find a (good) solution for this problem,
I decided to code myself.

https://github.com/Sordit/TimoBox

I created a Chrome extension, that will be installed on the Raspberry in the Chromium browser.
On the raspberry a Node.JS server is started in a docker container that will controll the extension and over the extension the Spotify WebPlayer.

It works exacltly the same as the Phoniebox with NFC scanner. (Extension will read the input and send it to the server)

Good thing: Spotify cannot break it, unless they disable the WebPlayer. :-)

But it may not be what you want, as it is maybe not 100% inside the ToS of Spotify. But at least for my niece and my girl it will work. ;-)

This is just working and I may break things with coming changes, while I teak it and add new features.

Have Fun testing and please give me feedback.

**Comment by DavidOrtmann** [comment]
I think it's an ingenious idea and I can't wait to try it out. However,
when opening open.spotify.com, I get the message "Payback of protected
content is not enabled". What can I do from there?

Cheers,
David

Sordit ***@***.***> schrieb am Sa., 4. März 2023, 13:27:

**Reply by Sordit** [reply]
Yes sorry. I forgot to add to the Readme that you need
sudo apt-get update
sudo apt-get install libwidevinecdm0

And I guess a reboot is needed as well
to enable DRM media

**Comment by DonPavlov** [comment]
I am kind of sad ,that no current conclusive answer has been reached. 
Yes reverse engineering the spotify protocol is kind of not what they allow.
But projects like spotifyd still exist and can easily be used. 
If the only thing i need to do is add in my own developer account and add my own app, i am fine doing that.

My question is: 
Will this community either try to use mopidy with its current spotify implementation(some kind of unreleased beta version of it), 
or will we switch over to something else. 

Once a decision has been reached we can focus our development effort on that. No big deal. 

I even want to help out myself, because it takes too long for my taste to reach any kind of concensus.

**Comment by Hamsert** [comment]
https://github.com/MiczFlor/RPi-Jukebox-RFID/issues/1815#issuecomment-1666535983
Following these instructions on a new bullseye installation made spotify usable again like it used to be! Even my old cards still work!

**Comment by nanu-c** [comment]
I have a working spotify installation with a raspberry pi by using https://spotifyd.rs/ and a python script written around the https://spotipy.readthedocs.io/en/2.22.1/.

```
#!/usr/bin/env python
#coding: utf8 
import RPi.GPIO as GPIO
import time
import RPi.GPIO as GPIO
import time
import spotipy
from spotipy.oauth2 import SpotifyOAuth

import logging
import sys

logger = logging.getLogger()
logger.setLevel(logging.INFO)
formatter = logging.Formatter('%(asctime)s | %(levelname)s | %(message)s')

stdout_handler = logging.StreamHandler(sys.stdout)
stdout_handler.setLevel(logging.DEBUG)
stdout_handler.setFormatter(formatter)

file_handler = logging.FileHandler('logs.log')
file_handler.setLevel(logging.DEBUG)
file_handler.setFormatter(formatter)


logger.addHandler(file_handler)
logger.addHandler(stdout_handler)

# Zählweise der Pins festlegen
GPIO.setmode(GPIO.BCM)
# Pin 18 (GPIO 24) als Eingang festlegen
GPIO.setup(4, GPIO.IN, pull_up_down = GPIO.PUD_UP)
GPIO.setup(27, GPIO.IN, pull_up_down = GPIO.PUD_UP)
GPIO.setup(22, GPIO.IN, pull_up_down = GPIO.PUD_UP)
GPIO.setup(23, GPIO.IN, pull_up_down = GPIO.PUD_UP)
GPIO.setup(24, GPIO.IN, pull_up_down = GPIO.PUD_UP)
i=0

# see https://spotipy.readthedocs.io/en/2.22.1/#client-credentials-flow
client_id = ""
client_secret = ""
redirect_uri = ""
scope = "user-read-playback-state,user-modify-playback-state"

sp = spotipy.Spotify(
        auth_manager=spotipy.SpotifyOAuth(
          client_id=client_id,
          client_secret=client_secret,
          redirect_uri=redirect_uri,
          scope=scope, open_browser=False))

# Shows playing devices
res = sp.devices()
volume = 0
deviceId = -1
def detectDevice():
    global deviceId
    global volume
    res = sp.devices()
    for device in res['devices']:
        if device['name'] == 'diele_musikbox':
            logging.info("found self")
            deviceId = device["id"]
            volume = device['volume_percent']
            if device['is_active'] == False:
                logging.info("transfer")
                sp.transfer_playback(device_id=device['id'])
            # sp.transfer_playback(device_id=device['id'], force_play=True)
            break
    logging.info(res)
    logging.info("volume: " + str(volume))
    logging.info("deviceId: " + str(deviceId))
detectDevice()
# Change track

player_is_spotifyd = True

def nextSong(_):
    logging.info("next song")
    if player_is_spotifyd:
        sp.next_track()

def previousSong(_):
    logging.info("previous song")
    if player_is_spotifyd:
        sp.previous_track()
isPlaying = False

def playPause(_):
    global isPlaying
    if isPlaying:
        isPlaying = False
        logging.info("pause")
        if player_is_spotifyd:
            sp.pause_playback()

    else:
        isPlaying = True
        logging.info("play")
        if player_is_spotifyd:
            detectDevice()
            # sp.transfer_playback(device_id=deviceId, force_play=True)
            sp.start_playback(device_id=deviceId)


def volumeUp(_):
    global volume
    volume = volume + 5
    logging.info("volume up")

    if volume > 100:
        volume = 100
    if player_is_spotifyd:
        sp.volume(volume)

def volumeDown(_):
    global volume
    volume = volume - 5
    logging.info("volume down")
    if volume < 10:
        volume = 10
    if player_is_spotifyd:
        sp.volume(volume)




GPIO.add_event_detect(4, GPIO.FALLING, callback = volumeDown, bouncetime = 700)
GPIO.add_event_detect(27, GPIO.RISING, callback = nextSong, bouncetime = 700)
GPIO.add_event_detect(22, GPIO.RISING, callback = playPause, bouncetime = 700)
GPIO.add_event_detect(23, GPIO.RISING, callback = previousSong, bouncetime = 700)
GPIO.add_event_detect(24, GPIO.RISING, callback = volumeUp, bouncetime = 700)


while 1:
    time.sleep(0.1)
```
